### PR TITLE
Instantiate classdumper

### LINKED ISSUE #208: StaticLinkageChecker to have instance variable for class repository
This is to avoid passing class loader and SyntheticRepository around static methods.

Followup for @netdpb 's comment https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/187/files#r230416247

### LINKED ISSUE #209: Class to take option for StaticLinkageChecker
By having a class to represent (command line) option, we can simplify `StaticLinkageChecker.getInstanceFromArguments`.

### PR DESCRIPTION
**Pull request by suztomo**
@netdpb This is a followup PR for your previous review. Would you check them?

- ClassDumper as an instance variable of StaticLinkageCheck
  Fixes #208. This eliminates code passing around jar file lists, BCEL SyntheticRepository and ClassLoader the among static methods.
- StaticLinkageCheckOption for command line option parsing.
  Fixes #209 
- use of immutable classes
- removal of unused methods that take InputStream as input

Out of scope:
The enhancement for the separating data collection from jar files and usage traversal (#203) is out of scope of this PR.

### CODE REVIEW

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java [hunk 1]
```diff
@@ -44,90 +48,21 @@ public void cleanup() throws IOException {
     classFileInputStream.close();
   }
 
-  @Test
-  public void testListConstantPool() throws IOException {
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java]: These methods that take InputStream were not used.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        "report-only-reachable",
+        false,
+        "To report only linkage errors reachable from entry point");
+
+    CommandLineParser parser = new DefaultParser();
+    List<Path> jarFilePaths = new ArrayList<>();
+
+    ImmutableList.Builder<String> mavenCoordinates = ImmutableList.builder();
+    try {
+      CommandLine cmd = parser.parse(options, arguments);
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: cmd --> command (Google style is not to abbreviate variables)

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+      boolean reportOnlyReachable = cmd.hasOption("r");
+
+      return builder()
+          .setBomCoordinate(mavenBomCoordinate)
+          .setMavenCoordinates(mavenCoordinates.build())
+          .setJarFileList(jarFilePaths)
+          .setReportOnlyReachable(reportOnlyReachable)
+          .build();
+    } catch (ParseException ex) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: Does it make sense to just let the ParseException be thrown instead?

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java [hunk 4]
```diff
@@ -142,4 +77,57 @@ public void testMethodDescriptorToClass_primitiveTypes() {
...
+        "com.google.firestore.v1beta1.FirestoreGrpc$FirestoreImplBase",
+        "com.google.firestore.v1beta1.FirestoreGrpc$FirestoreFileDescriptorSupplier"
+    );
+  }
+
+  @Test
+  public void testCreationInvalidInput() {
+    try {
+      ClassDumper.create(ImmutableList.of(Paths.get("")));
+      Assert.fail("Empty path should generate IOException");
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java]: but you're catching a RuntimeException, not an IOException?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: Fixed

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: I added helpFormetter.printHelp after catching ParseException.
Now ParseException is re-thrown.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java (hunk 4 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassDumperTest.java]: Updated message

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setBomCoordinate(@Nullable String value);
+    abstract Builder setMavenCoordinates(List<String> value);
+    abstract Builder setJarFileList(List<Path> value);
+    abstract Builder setReportOnlyReachable(boolean value);
+    abstract StaticLinkageCheckOption build();
+  }
+
+  static StaticLinkageCheckOption parseArgument(String[] arguments) throws ParseException {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: parseArguments

**Comment by suztomo** [comment]
@netdpb I'll incorporate your comment if you have any in next PR. Merging the branch for now.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java (hunk 5 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java]: Missed this. I'll update the function name in next PR.

**Pull request merged**